File: c_tests/mandle.c
#include <cstdio>
#include <stdio.h>

template <typename T> void mandle()
{
    const T cx = 0.0458;
    const T cy = 0.08333;
    const T c2 = 2.0;
    const T c4 = 4.0;

    for ( int y = -12; y <= 12; y++ )
    {
        for ( int x = -39; x <= 39; x++ )
        {
            const T ca = x * cx;
            const T cb = y * cy;
            T a = ca;
            T b = cb;
            int i;
            for ( i = 0; i <= 15; i++ )
            {
                T t = a * a - b * b + ca;
                b = c2 * a * b + cb;
                a = t;
                if ( ( a * a + b * b ) > c4 )
                    break;
            }
            if ( i > 15 )
                printf( " " );
            else
            {
                if ( i > 9 )
                    i = i + 7;
                printf( "%c", 48 + i );
            }
        }
        printf("\n");
    }
} //mandle

int main()
{
    mandle<float>();
    mandle<double>();
    mandle<long double>();

    printf( "mandle completed with great success\n" );
    return 0;
} //main

File: pascal_tests/mandle.pas
program mandle;

procedure mandle_single;
var
  cx, cy, c2, c4: Single;
  ca, cb, a, b, t: Single;
  x, y, i: Integer;
begin
  cx := 0.0458;
  cy := 0.08333;
  c2 := 2.0;
  c4 := 4.0;

  for y := -12 to 12 do
  begin
    for x := -39 to 39 do
    begin
      ca := x * cx;
      cb := y * cy;
      a := ca;
      b := cb;
      i := 0;
      while i <= 15 do
      begin
        t := a * a - b * b + ca;
        b := c2 * a * b + cb;
        a := t;
        if (a * a + b * b) > c4 then
          break;
        Inc(i);
      end;
      if i > 15 then
        Write(' ')
      else
      begin
        if i > 9 then
          i := i + 7;
        Write(Chr(48 + i));
      end;
    end;
    WriteLn;
  end;
end;

procedure mandle_double;
var
  cx, cy, c2, c4: Double;
  ca, cb, a, b, t: Double;
  x, y, i: Integer;
begin
  cx := 0.0458;
  cy := 0.08333;
  c2 := 2.0;
  c4 := 4.0;

  for y := -12 to 12 do
  begin
    for x := -39 to 39 do
    begin
      ca := x * cx;
      cb := y * cy;
      a := ca;
      b := cb;
      i := 0;
      while i <= 15 do
      begin
        t := a * a - b * b + ca;
        b := c2 * a * b + cb;
        a := t;
        if (a * a + b * b) > c4 then
          break;
        Inc(i);
      end;
      if i > 15 then
        Write(' ')
      else
      begin
        if i > 9 then
          i := i + 7;
        Write(Chr(48 + i));
      end;
    end;
    WriteLn;
  end;
end;

procedure mandle_extended;
var
  cx, cy, c2, c4: Extended;
  ca, cb, a, b, t: Extended;
  x, y, i: Integer;
begin
  cx := 0.0458;
  cy := 0.08333;
  c2 := 2.0;
  c4 := 4.0;

  for y := -12 to 12 do
  begin
    for x := -39 to 39 do
    begin
      ca := x * cx;
      cb := y * cy;
      a := ca;
      b := cb;
      i := 0;
      while i <= 15 do
      begin
        t := a * a - b * b + ca;
        b := c2 * a * b + cb;
        a := t;
        if (a * a + b * b) > c4 then
          break;
        Inc(i);
      end;
      if i > 15 then
        Write(' ')
      else
      begin
        if i > 9 then
          i := i + 7;
        Write(Chr(48 + i));
      end;
    end;
    WriteLn;
  end;
end;

begin
  mandle_single;
  mandle_double;
  mandle_extended;

  WriteLn('mandle completed with great success');
end.
